tchannel/tornado/connection.py

```diff
@@ -452,12 +466,26 @@ def serve(self, handler):
         """
         assert handler, "handler is required"
         assert self._loop_running, "Finish the handshake first"
-
         while not self.closed:
             message = yield self.await()
-
             try:
-                handler(message, self)
+                if self.draining:
```

Comment:
Right now you're doing high-level logic inside low-level connection logic. The dispatcher is the place where we should be denying inbound requests. You can do this while also keeping the registered handlers.

- Put dispatcher in a state where it starts returning declines
- Call close on all of the peers we want to kill
- Connection makes a note of its outstanding requests/responses and returns a future that resolves when all of those are done
- A timeout callback actually kills the connection after some period of time